### DISCUSSION TITLE
Upcoming September 2022 Features

### DISCUSSION THREAD

**Discussion by johnlindquist** [johnlindquist/kit#868] [General]
The next Script Kit release is right around the corner. We've been putting a lot of focus into the user experience and cleaning up the UI. Below are the features you can expect to see in the upcoming build. If you'd like to help test it out (expect minor bugs!), join us over on the new Discord server: 

https://discord.gg/qnpVwbPy

## Keyboard Shortcuts and Buttons

The `shortcuts` array allows you to define alternate behaviors for your prompt that are accessible and presented as keyboard shortcuts at the bottom of the prompt. You can also click the buttons if needed. 

In previous releases, you could just hit "Enter" to submit, now you can do _anything_ 🤯. If you have any shortcut ideas you want to explore, please comment below and I'm definitely happy to help out:

> Note: `flags` and flag shortcuts still work and are accessible through the right arrow as always. They've just always been "hidden" and shortcuts help users better understand what's important to the script

<img width="830" alt="CleanShot 2022-08-20 at 07 54 43@2x" src="https://user-images.githubusercontent.com/36073/185749458-d846a521-bc14-461b-81f2-4c244dc889e6.png">

```js
// Name: Shortcuts Example

import "@johnlindquist/kit"

let clearInput = {
            name: "Clear",
            key: "cmd+u",            
            onPress: async () => { 
                setInput("")
            },
            // optional, but will only display when set
            bar: "right",
}

let reloadChoices = {
            name: "Reload",
            key: "cmd+l",
            bar: "right",
            onPress: async () => { 
                setChoices(["one","two","three"])
            }
}      

let submitInputNotChoice = {
            name: "Submit Input",
            key: "cmd+i",
            bar: "right",
            onPress: async (input) => { 
                submit(input)
            }
}                

let runAppLauncher = {
            name: "Launch App",
            key: "cmd+a",
            bar: "left",
            onPress: async () => { 
                await run(kitPath("main", "app-launcher.js"))
            }
}

let result = await arg({
    placeholder: "Shortcut demo",
    shortcuts: [        
        clearInput,
        reloadChoices,
        submitInputNotChoice,
        runAppLauncher
    ]
}, ["apple", "banana", "cherry"])

await div(md(`## ${result}`))
```

[Open shortcuts-example in Script Kit](https://scriptkit.com/api/new?name=shortcuts-example&url=https://gist.githubusercontent.com/johnlindquist/284da4c495d497f4b61be86baa4f3e9b/raw/0f618f611bcb514b1cda03ed722f2bc38efc2ab3/shortcuts-example.js")

## Fields Prompt

`await fields()` is now the quickest way to get multiple strings values from the user. The main goal of `fields` is simplicity and speed. Type a couple of strings, then hit `Enter`.

If you need a more robust feature set (autocomplete, previews, etc) then `await arg()` should have you covered.

```js
// Name: Fields Example

import "@johnlindquist/kit"

// Basic strings
let [first, last] = await fields("First Name", "Last Name")

// Objects with input attributes
let [date, amount] = await fields({
    label: "When will you be there?",
    type: "date"
}, {
    label: "How many apples?",
    type: "number",
    min: 1,
    max: 100,
    value: "50"
})


await div(md(`# ${first} ${last}

* ${date}
* ${amount}

`))
```

[Open fields-example in Script Kit](https://scriptkit.com/api/new?name=fields-example&url=https://gist.githubusercontent.com/johnlindquist/ffd40b12fe727ea0331dc6f4c8954b9e/raw/89119030fbd9a888610be330386b26cef3ee3825/fields-example.js")

## `enter`

Defining `enter` helps the user know what's happening next. This was a _major_ reason for this update as in all of my one-on-one sessions with Script Kit users, everyone was a bit hesitant to hit then Enter key when they didn't know what was going to happen:

<img width="459" alt="CleanShot 2022-08-20 at 07 59 11@2x" src="https://user-images.githubusercontent.com/36073/185749716-fba0d435-ad5a-49a2-b657-b6952832910e.png">

```js
// Name: Enter Example

import "@johnlindquist/kit"

let username = await arg({
    placeholder: "Type your username",
    resize: true,
    shortcuts:[],
    // Tell the user what will happen "next"
    enter: "Submit to Github API"
})

// Do stuff with username
```

[Open enter-example in Script Kit](https://scriptkit.com/api/new?name=enter-example&url=https://gist.githubusercontent.com/johnlindquist/ac373c9cad03a32daf1338a379f2448d/raw/0f0fafbe141e87c8f4dc6dc343014e399da76a87/enter-example.js")

## Toggle Between Editing and Running a Script

When running a script, press `cmd+o` to open the script in your editor (unless you override `cmd+o` in your shortcuts, it will be auto-added, even if the button isn't visible). From the Kit editor, pressing `cmd+shift+s` will save and run the script. This makes it _SUPER FAST_ to jump between editing and running contexts and really boosts your script editing workflow:

(I use the mouse and buttons in the video to explicitly show what's happening. It's way cooler to use the keyboard shortcuts 😎)

https://user-images.githubusercontent.com/36073/185750746-9c8339a2-9f59-4735-bf1d-b70d624bc2dc.mp4

## `resize`

After building the `fields` prompt, I recognized there was too much empty space in simple prompts, so I began implementing an automatically resizing window. `arg` is a special case because it can "lazy load" choices/previews/etc which makes the prompt impossible to measure right when it opens. So when you're using `arg` and want it to resize, set `resize` to `true`. On the flip side, you can set `resize` to `false` for other components if you don't want the resize behavior:

> Note: `cmd+0`(that's a zero) from the main menu will reset all prompt positions in case something resizing/re-positioned improperly

<img width="448" alt="CleanShot 2022-08-20 at 08 08 49@2x" src="https://user-images.githubusercontent.com/36073/185750265-b827ad34-eff3-4fb3-a91e-9a6ffae980e9.png">

```js
// Name: Resize Example

import "@johnlindquist/kit"

await arg({
    resize: true
})
```

[Open resize-example in Script Kit](https://scriptkit.com/api/new?name=resize-example&url=https://gist.githubusercontent.com/johnlindquist/c7ba94e870feacba4aa9f6db26f2f7a0/raw/39e47f0b307c4e789a6f145fd80a8175aa44bf25/resize-example.js")

## Dev Settings in the Menubar

Because we're developers, we want access to _all_ the settings. I'm attempting to expose as much as possible in the menubar so you can dig in and mess with whatever you need:

<img width="768" alt="CleanShot 2022-08-20 at 08 28 41@2x" src="https://user-images.githubusercontent.com/36073/185751420-c20f0f23-aff4-4e77-ad91-847c47507927.png">

**Comment by dadamssg** [comment]
seems the discord link is invalid..?

![image](https://user-images.githubusercontent.com/881986/188924907-8f616b00-f978-4005-baba-64fe3a2fafd5.png)

**Reply by johnlindquist** [reply]
@dadamssg They're set to expire after a week. Here's a new one:

https://discord.gg/uPBu2Fg7

**Comment by pomdtr** [comment]
Hey @johnlindquist, Could you send me an invite to the discord server ?
I'm working on a command line launcher with a plugin api based on stdin/stdout, and I would love to integrate it with scriptkit.

Sneak peak:

https://user-images.githubusercontent.com/17577332/191946507-d17e93aa-0f0c-4038-877c-1733d56575b1.mp4

**Reply by johnlindquist** [reply]
Looks awesome @pomdtr !

Here you go: https://discord.gg/Sh5BNTEs

I'm afk for a few hours if I don't respond right away 😇

**Reply by pomdtr** [reply]
No problem, I'm usually working during office hour 😅 

I would love to demo it to you, I have a lot of ideas which am quite proud of api-wise.
